packages/connection/index.js

```diff
@@ -106,12 +107,25 @@ class Connection extends EventEmitter {
     return this.jid
   }
 
-  _online() {
-    this.emit('online', this.jid)
+  _status(status, ...args) {
+    this.status = status
+    this.emit('status', status, ...args)
+    this.emit(status, ...args)
   }
 
-  _authenticated() {
-    this.emit('authenticated')
+  ready(fn) {
+    if (fn) {
+      if (this.status === 'online') {
+        return Promise.resolve().then(fn)
+      }
+      this.on('online', fn)
+      return
+    }
+
+    if (this.status === 'online') {
+      return Promise.resolve()
+    }
+    return this.once('online')
```

Comment:
not sure the promise interface is useful

```js
entity.ready().then(() => {
...
})
```

will only work for the first connection, not for reconnection

probably remove this, in the rare case someone needs to do something only on first connection, use `status === 'online' || once('online')`